torchtitan/parallelisms/pipelining_utils.py

```diff
@@ -9,10 +9,61 @@
     ScheduleFlexibleInterleaved1F1B,
     ScheduleInterleaved1F1B,
 )
-from torch.distributed.pipelining.schedules import get_schedule_class
+
+from torch.distributed.pipelining.schedules import (
+    get_schedule_class,
+    PipelineScheduleMulti,
+    PipelineScheduleSingle,
+)
 from torchtitan.logging import logger
 
 
+def generate_split_points(job_config, pp_dim, model_config):
+    schedule_class = get_schedule_class(
+        job_config.experimental.pipeline_parallel_schedule
+    )
+    if issubclass(schedule_class, PipelineScheduleSingle):
+        num_stages_per_rank = 1
+    elif issubclass(schedule_class, PipelineScheduleMulti):
+        # Multi-stage schedules support more than 2 stages per rank, but this is the default if
+        # no pipeline split is specified
+        num_stages_per_rank = 2
+    else:
+        raise ValueError(
+            f"Unsupported pipeline schedule: {job_config.experimental.pipeline_parallel_schedule}"
+        )
+    total_stages = pp_dim * num_stages_per_rank
+    num_layers = model_config.n_layers
+    if total_stages > num_layers:
+        raise ValueError("Total stages cannot be greater than the number of layers")
+
+    base_interval = num_layers // total_stages
+    remainder = num_layers % total_stages
+
+    # Calculate the number of extra layers to add to the middle stages
+    extra_layers = min(remainder, max(0, total_stages - 2))
```

Comment:
Upon first read, I have the feeling that there is no need for this `extra_layers`, because:
There are two cases
1. `extra_layers = remainder`. In this case we can just use `remainder`.
2. `remainder = total_stages - 1` and `extra_layers = total_stages - 2 = remainder - 1`. In this case even if you use `remainder` instead of `extra_layers`, the code will do the same thing -- put one extra layer for each middle stage, and put the last remaining layer to the last stage, with `remainder = 1` after the code finishes.

Btw `total_stages` must be >= 2 as `pp_dim >= 2` and `num_stages_per_rank >= 1` so no need for the max here. We can add an assertion if you think it's needed.

Reply:
Updated! I just simplified it